### DISCUSSION TITLE
Help needed with HTMLCompletionProvider and HtmlTypedTextInterceptor

### DISCUSSION THREAD

**Discussion by Chris2011** [apache/netbeans#4723] [Q&A]
I'm working on a feature to automatically close html tags as you can see it here in

VS Code:
![vscode-auto-end-tag-insertion](https://user-images.githubusercontent.com/795658/193478437-7f85078a-6f2a-4b79-879c-932b494c85e1.gif)

or in WebStorm:
![webstorm-auto-end-tag-insertion](https://user-images.githubusercontent.com/795658/193478441-a49593b6-ff71-4e73-a715-02645de0e844.gif)

So right after I hit ">" I want to insert the closing tag of the corresponding open tag e.g. \<div">\</div>" In general, it was easy, I found the part in the [HtmlTypedTextInterceptor](https://github.com/apache/netbeans/blob/master/ide/html.editor/src/org/netbeans/modules/html/editor/typinghooks/HtmlTypedTextInterceptor.java#L111). So I added a function, check for opening tag and automatically add the end tag. So far so good. This works for those examples:

```html
<div> -> <div>|</div>
<div class=""> -> <div class="">|</div>
```

There is, atm just one problem where the inteceptor, not intercept: `<div class="t"> -> <div class="t"> (Code completion comes up)`

So my problem is that the difference between `<div class=""> and <div class="t">` is the value inside the quotes of the class attribute. I jumped in with the debugger and I can't figure out why the Interceptor doesn't jump in here. With the value, the [HtmlCompletionProvider](https://github.com/apache/netbeans/blob/master/ide/html.editor/src/org/netbeans/modules/html/editor/completion/HtmlCompletionProvider.java#L374) jumps in.

Can someone please give me a hint where to find when the HtmlCompletionProvier jumps in and when HtmlTypedTextInterceptor? I need to check what's going on with the value and trying to fix this.

**Comment by matthiasblaesing** [comment]
You are hitting a border case. There are two different mime types at play here:

- ` text/html`  (outside)
- ` text/css`  (the "t" token behind class)

You can observe this, if you put a breakpoint in `org.netbeans.modules.editor.lib2.typinghooks.TypedTextInterceptorsManager#getInterceptors`  (current master line 170 for example). Then  type the ``>` after `"t"`. You observe, that the reported `mimePath`  is ` text/html/text/css`.

**Reply by Chris2011** [reply]
Great, will dive into it. Thx for the hint :)
